### ISSUE TITLE
How do I create a custom neural network in TensorFlow and run as an external brain?

### ISSUE THREAD

**Issue by RyanRizzo96** [Unity-Technologies/ml-agents#696]
Hi everyone,

I have completed all the necessary steps to install ML-Agents on Windows 10 following this excellent tutorial: 

https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Installation-Windows.md 👍

Next I followed this [comprehensive guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Getting-Started-with-Balance-Ball.md) to get started with the balance ball project. _All has worked well._

Now, I want to go a step further and create not just my own new scenes and environments but also **create the neural networks in Python**. Since v 0.3 has done away with the PPO.ipynb file which I could edit and run the code in Jupyter notebook, I am stuck as to where I can actually code my custom neural network using Python and TensorFlow to use this as an external brain to my agent.

1. 1. Is it possible to create a **custom neural network** using Python and TensorFlow and somehow attach this as an external brain to my agent?
2. 2. Which file should I edit to start coding my custom neural network? 
3. Is this necessary from my side to build a neural network from scratch, or does unity ML-Agents provide all the coverage needed?
1. Is it possible to create a **custom neural network** using Python and TensorFlow and somehow attach this as an external brain to my agent?
2. Which file should I edit to start coding my custom neural network? 
3. Is this necessary from my side to build a neural network from scratch, or does unity ML-Agents provide all the coverage needed?

To give you a broader view of what I am trying to accomplish. I have a small robotic system modeled in unity to include friction coefficients and servo motor PID tuning to mimic a real-life robotic system. I now want to train this system using TensorFlow and Python by coding up a recurrent neural network. I was wondering whether this would be the way to go, since machine learning in Python and TensorFlow is 'easier' than in C#. 

I cam across the [Training ML-Agents documentation](https://github.com/Unity-Technologies/ml-agents/blob/bbbe2e7e8f5b87d62650d14c9a21e45f1a566596/docs/Training-ML-Agents.md#training-config-file), but I am not talking about simply editing the config file `training_config.yaml`, but rather building a neural network from scratch.

In short I am looking to alter the Python API.

![image](https://user-images.githubusercontent.com/31866965/39396666-b45c11a2-4af2-11e8-8bd4-37cf1698a1e1.png)

I could not seem to find any information regarding the above two points and have been stuck trying to figure this out. Any help would be greatly appreciated. 

Thanks!

**Comment by joaquines** [comment]
yes you can. all the python code that creates the TF layers is included. you do not need to touch it for it to work, but if might be a good place to start if you want a custom topology.
Take a look at https://github.com/Unity-Technologies/ml-agents/blob/master/python/unitytrainers/models.py#L162

**Comment by mmattar** [comment]
Hi @joaquines - are you still facing any issues?

**Comment by xiaomaogy** [comment]
Thanks for reaching out to us. Hopefully you were able to resolve your issue.  We are closing this due to inactivity, but if you need additional assistance, feel free to reopen the issue.

**Comment by UgurUysal86** [comment]
Is it possible to use keras-rl in combination with ML-agents? I would like to create custom networks. Thank you.

**Comment by xiaomaogy** [comment]
@UgurUysal86 You can of course do write your own trainer code and not use our trainer code at all.

**Comment by UgurUysal86** [comment]
Thank you! Is there any guide on how to create your own trainer?

**Comment by xiaomaogy** [comment]
Emm....@UgurUysal86 if you don't know how to do that, then a guide won't help you much.

**Comment by UgurUysal86** [comment]
Ouch. Thank you anyway.

**Comment by MarkTension** [comment]
> yes you can. all the python code that creates the TF layers is included. you do not need to touch it for it to work, but if might be a good place to start if you want a custom topology.
> Take a look at https://github.com/Unity-Technologies/ml-agents/blob/master/python/unitytrainers/models.py#L162

This link would be helpful to me, but is broken, any new address to get here? Thanks!

**Comment by iamPres** [comment]
> > yes you can. all the python code that creates the TF layers is included. you do not need to touch it for it to work, but if might be a good place to start if you want a custom topology.
> > Take a look at https://github.com/Unity-Technologies/ml-agents/blob/master/python/unitytrainers/models.py#L162
> 
> This link would be helpful to me, but is broken, any new address to get here? Thanks!
Pretty sure this is it: https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/models.py

I am trying to figure this out as well. Please let me know if you figure anything out there is very little documentation

**Comment by iamPres** [comment]
> > yes you can. all the python code that creates the TF layers is included. you do not need to touch it for it to work, but if might be a good place to start if you want a custom topology.
> > Take a look at https://github.com/Unity-Technologies/ml-agents/blob/master/python/unitytrainers/models.py#L162
> 
> This link would be helpful to me, but is broken, any new address to get here? Thanks!

**Here you go: https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/models.py**

Please tell me if you figure anything out. I'm trying to do the same thing.****

**Comment by MarkTension** [comment]
Thanks! Will get back at you when I get it working for me

**Comment by iamPres** [comment]
> Thanks! Will get back at you when I get it working for me
I am also trying to model a robotic system using Unity, but found it very challenging to model friction and elasticity coefficients. Instead, I'm training a network in Unity to find target positions, and then using RL in real life to teach the robot how to get to said positions, so that it can learn to perform with friction and elasticity. Do you have any thoughts on this approach? 

Thanks

**Comment by iamPres** [comment]
> Thanks! Will get back at you when I get it working for me

I just got it working. Since there isn't any documentation for the latest version, I downgraded to unity 2018.1.0. From there, I just forllowed [https://medium.com/@indiecontessa/setting-up-a-python-environment-with-tensorflow-on-macos-for-training-unity-ml-agents-faf19d71201](this medium article) and created my own academy, brain, and agent. I also created my own script to interface with the environment python/genetic_algorithm.py since I don't want to use RL. 

Here is all my code:
https://github.com/iamPres/Bipedal-Walker

Hope this helps
